### PR TITLE
OAuthbearer support

### PR DESCRIPTION
**Pull request by macasado86**
Hi!

I have opened this PR to include the necessary functionality to authenticate Erlkaf consumers and producers using the OAuthbearer protocol:

- OIDC authentication passing the required parameters to librdkafka: `sasl_oauthbearer_method, sasl_oauthbearer_client_id, sasl_oauthbearer_client_secret, sasl_oauthbearer_scope sasl.oauthbearer.extensions, sasl_oauthbearer_token_endpoint_url`

- Custom refresh token callback to provide users a way to implement their retrieve access token function and pass it to librdkafka

I have needed to enable this options in order to retrieve the first token before trying to establish the connection with the brokers: 
  - rd_kafka_sasl_background_callbacks_enable
  - rd_kafka_conf_enable_sasl_queue

This way, the refresh callbacks are triggered automatically by librdkafka in a background thread.

Any suggestions or changes you think I should make, please let me know.

### CODE REVIEW

**Comment by silviucpp** [comment]
Hello,

The implementation looks ok. But please fix the duplicate code issues I mentioned in the review.

**Comment by macasado86** [comment]
Hi!

I can't see the review. Maybe you forgot to submit it.

Thanks

**Comment by silviucpp** [comment]
You cannot see the above change requests ? 
Screenshot 2024-07-24 at 11 10 28

**Comment by macasado86** [comment]
No, I can't. I think you have to submit the review (it's in 'pending' state)

**Comment by silviucpp** [comment]
Now you can see it ?

File: c_src/erlkaf_producer.cc [hunk 1]
```diff
@@ -417,3 +444,105 @@ ERL_NIF_TERM enif_produce(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 
     return ATOMS.atomOk;
 }
+
+char** split_producer_extensions(const std::string extensions_str, size_t* length)
```
[the remark this answers is not in the document]
Comment (macasado86) [reply c_src/erlkaf_producer.cc]: I have moved this function to a new file called `erlkaf_oauthbearer.cc`

File: c_src/erlkaf_producer.cc [hunk 2]
```diff
@@ -417,3 +444,105 @@ ERL_NIF_TERM enif_produce(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+    for(size_t i = 0; i < *length; ++i)
+    {
+        extensions[i] = new char[extensions_vector[i].size() + 1];
+        strcpy(extensions[i], extensions_vector[i].c_str());
+    }
+
+    return extensions;
+}
+
+void free_producer_extensions(char** extensions, size_t length)
```
[the remark this answers is not in the document]
Comment (macasado86) [reply c_src/erlkaf_producer.cc]: I have moved this function to a new file called `erlkaf_oauthbearer.cc`

File: c_src/erlkaf_producer.cc [hunk 3]
```diff
@@ -417,3 +444,105 @@ ERL_NIF_TERM enif_produce(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+            return ATOMS.atomError;
+        }
+        else
+        {
+            free_producer_extensions(extension_key_value, extension_key_value_cnt);
+            return ATOMS.atomOk;
+        }
+}
+
+ERL_NIF_TERM enif_producer_oauthbearer_set_token_failure(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
```
[the remark this answers is not in the document]
Comment (macasado86) [reply c_src/erlkaf_producer.cc:479]: I have moved some of this functionality (all but parsing the arguments) to a new file called `erlkaf_oauthbearer.cc`

**Pull request merged**